### ISSUE TITLE
(tests only) Add a "simple" mode

### ISSUE THREAD

**Issue by jackh726** [rust-lang/chalk#299]
So, when solving a goal, `chalk-solve` automatically adds various clauses for potential downstream impls, `Compatible`, etc. Additionally, any tests not in `test/slg` also are checked that they don't violate orhpan rules, etc. While, this is definitely good in general, it makes actually debugging of logic (using `CHALK_DEBUG=1` or `2` much more noisy.

I think it would be super helpful to somehow have a mode for tests where only the clauses specified by the test are used and the program isn't checked. Importantly, this would be completely opt-in (so maybe an environment variable) and would never be used by default for any tests.

This would be pretty easy to implement:

This `checked_program` call would need to be conditional:
https://github.com/rust-lang/chalk/blob/3ef20308e7882a3e5fb71a6f28caab8cc71cf5f0/tests/test/mod.rs#L148

Also, the extra clauses get added in `program_clauses.rs`, for example:
https://github.com/rust-lang/chalk/blob/3ef20308e7882a3e5fb71a6f28caab8cc71cf5f0/chalk-solve/src/clauses/program_clauses.rs#L183
https://github.com/rust-lang/chalk/blob/3ef20308e7882a3e5fb71a6f28caab8cc71cf5f0/chalk-solve/src/clauses/program_clauses.rs#L415

Again, the purpose of this is to make debugging logic much easier. There will almost certainly be tests that fail with the option enabled, but that's okay because it isn't for them.

**Comment by Mcat12** [comment]
@jackh726 is this issue ready to go? It's labeled with "design" (aka "needs design work"), but seems like it is straightforward enough to start implementing.

Maybe, we need to decide if this is important or if it will be used. Do you envision this only being used during test creation or debugging? Also, would this option be per-test (add syntax to `test` macro) or for all tests (environment variable)?

**Comment by jackh726** [comment]
@Mcat12 yeah so I marked this a "design" more so because it "needs decision". It's actually pretty straightforward to implement.

(I wanted to just link a comment @nikomatsakis made in the last meeting, since it's relevant https://rust-lang.zulipchat.com/#narrow/stream/144729-wg-traits/topic/design.20meeting.202020.2E05.2E26/near/198818703)

The points you make are completely valid. When I first made this issue, I mostly thought that this might just be an option you toggle for debugging. (Based on @nikomatsakis's comments and my own thoughts, I'm not sure I completely feel that way anymore) But really, there's actually two separate "issues" here: 1) Implicit WF and coherence checks 2) "Extra" clauses added by chalk-solve (e.g. `Compatible, IsUpstream, CannotProve`)

The first is actually *pretty* straightforward for "yes, let's just not run them for most tests". It's just something to think about that we trade *automatic* checking of coherence and WF to *manual* checking when writing the tests. But again, this is also something that could easily be toggleable.

The second (extra clauses) is a bit more tricky. Since these do actually change how a goal is solved. So, I'm much more unsure about whether this is a good idea. Maybe, though, we could change how we *generate* these clauses in the first place. Since often times we generate these clauses when they aren't going to be useful. I also feel like are test coverage for actually using these is pretty poor. (At least, if you were to ask me which tests require them, I wouldn't be able to point you to them without digging in)

So, one counterpoint (in favor of *not* removing these), is I've definitely had times where I've made a change and tests fail in the coherence/WF checks. It's unclear if they would have *also* failed without them, but does suggest that if we do remove these implicit checks we should be careful that we have sufficient test coverage for what we lose. (This is somewhat moot if the simple mode is opt-in)

> Do you envision this only being used during test creation or debugging?

debugging for sure

> Also, would this option be per-test (add syntax to test macro) or for all tests (environment variable)?

Environment variable
